user_guide_src/source/changelogs/v4.3.0.rst

```diff
@@ -12,6 +12,9 @@ Release Date: Unreleased
 BREAKING
 ********
 
+- The data structure returned by ``BaseConnection::getForeignKeyData()`` has been changed.
+- The naming of foreign keys in OCI8 has changed. The previous suffix of ``_fk`` has changed to ``_foreign``.
```

Comment:
`_foreign` is a bit long. 
Unique key would be `_u` (See #6516).

We will make some breaking changes around the database code in 4.3. 
Why don't we change it to `_fk` instead?
@MGatner @michalsn @sclubricants

Reply:
I would favor _fk. However I would prefer not to make any guarantees.

Reply:
> What do you mean by guarantees? 
I mean provide documentation that assures this is the name they are going to get. All DBMS have limits on the length of names so it may not always be possible to provide such name. 

Its best practice for the developer to lookup the constraint names on a table and use the names provided by lookup.. unless they have set the name themselves.

>If the CONSTRAINT symbol clause is not defined, or a symbol is not included following the CONSTRAINT keyword:
>For InnoDB tables, a constraint name is generated automatically.
>For NDB tables, the FOREIGN KEY index_name value is used, if defined. Otherwise, a constraint name is generated automatically.

https://dev.mysql.com/doc/refman/5.6/en/create-table-foreign-keys.html#foreign-key-identifiers

A name isn't even required. The generated names are just a unique random string.

I think your too hung up on the names.

https://stackoverflow.com/questions/2240929/what-is-a-proper-naming-convention-for-mysql-fks

Reply:
>So, to specify a named constraint, use the key word CONSTRAINT followed by an identifier followed by the constraint definition. (If you don't specify a constraint name in this way, the system chooses a name for you.)

https://www.postgresql.org/docs/current/ddl-constraints.html

Reply:
Shorter seems bette to me. The changelog reads like it was already `_fk` though? See if @codeigniter4/database-team has any opinions.

Reply:
I prefer _fk. I use '_fk' to define foreign keys in my databases, never '_foreign'.

Reply:
In the current implementaion, only OCI8 uses `_fk`, others use `_foreign`.

Reply:
I run this on my db:

```SQL
SELECT
COUNT(DISTINCT t.TABLE_NAME) AS TABLE_COUNT,
ROUND(AVG(LENGTH(t.TABLE_NAME))) AS AVG_LENTH_TABLE,
MAX(LENGTH(t.TABLE_NAME)) AS MAX_LENGTH_TABLE,
c.COLUMN_COUNT,
c.AVG_LENTH_COLUMN,
c.MAX_LENGTH_COLUMN
FROM
information_schema.`TABLES` AS t
INNER JOIN (SELECT
            COUNT(DISTINCT CONCAT(t.TABLE_NAME,c.COLUMN_NAME)) AS COLUMN_COUNT,
            ROUND(AVG(LENGTH(c.COLUMN_NAME))) AS AVG_LENTH_COLUMN,
            MAX(LENGTH(c.COLUMN_NAME)) AS MAX_LENGTH_COLUMN
            FROM
            information_schema.`TABLES` AS t
            INNER JOIN information_schema.`COLUMNS` AS c ON c.TABLE_CATALOG = t.TABLE_CATALOG AND c.TABLE_SCHEMA = t.TABLE_SCHEMA AND c.TABLE_NAME = t.TABLE_NAME
            WHERE
            t.TABLE_TYPE NOT IN ('SYSTEM VIEW') AND
            t.TABLE_SCHEMA NOT IN ('mysql', 'performance_schema')) c
WHERE
t.TABLE_TYPE NOT IN ('SYSTEM VIEW') AND
t.TABLE_SCHEMA NOT IN ('mysql', 'performance_schema')
```
Results:

```
TABLE_COUNT       140
AVG_LENTH_TABLE   15
MAX_LENGTH_TABLE  37

COLUMN_COUNT      3132
AVG_LENTH_COLUMN  12
MAX_LENGTH_COLUMN 43
```

So:

```
prefix: 2
table:  15
column: 12
suffix: 2
------------
total:  31
```

So based on my averages I would fail the naming scheme even with the shorter `_fk` even with only a single column.

Using `_fk` would be better then `_foreign` but we still need to have an alternative when the name exceeds the maximum allowed.

Reply:
Changing `_foreign` to `_fk` is more tricky, since it may be quite a big problem for some use cases.

I would rather follow our current convention and change everything to `_foreign` - simply because it will generate fewer problems for developers.

But I won't be a blocker and if the majority of you guys think we should change it then I'm okay with it too. We just have to put it under the BC break list.

Reply:
Perhaps if we do change to _fk it would be better to allow manually setting the constraint name. This would make the transition easier because you could define the name as the old convention of _foreign.

Reply:
Yes, I would be in favor of enabling the possibility to set the constraint name manually. I believe it would simplify a lot of things.

Reply:
> Changing _foreign to _fk is more tricky, since it may be quite a big problem for some use cases.

Why tricky? What use cases will get quite a big problem?

Reply:
> Why tricky? What use cases will get quite a big problem?

With this change, we're forcing every existing project to change the constraint names. If I don't make a change in my current project, there are at least 2 scenarios where it might cause problems:

I have an existing application in production, and in my development environment, I fire the command:
```
php spark migrate:refresh
```
Now I'm screwed. Different constraints in production and development. Good luck to me with making changes now.

Second scenario. A new developer comes to the team, configures his environment from scratch, and now he has completely different constraint names in his dev and production environment. 

If this change would be necessary because of a bug or were introduced in v5, I wouldn't be opposed. But making such a change in a minor version is not responsible (from my perspective). It seems like we are making this change because we feel like it.

Reply:
If you didn't set the constraint name you probably ought not depend on a name. I would hope they would look up the constraint names in order to drop them.

The default behavior of most databases is to give them a short description with a unique code. This is if you don't define the name yourself.

Reply:
I agreed with michalsn. There is no compelling reason why the foreign key suffix change has to be done in the minor version.

Reply:
Well I have `_processForeignKeys()` all factored into a single method now.

Let me make the point though that as Forge is right now there is no way to add a foreign key to an existing table. So it would be less likely that anyone is adding/removing in migrations. If we add this ability which I intend to do after this PR then perhaps these sorts of migrations will become more common and therefore become a bigger headache later.

Another thing is use of foreign keys is fairly rare. I don't think I use any of them in all of my databases. I used them years ago but found them more of a headache than useful. This is probably one of the reasons we haven't heard of much complaint about the lack of features using foreign keys. Adding and removing foreign keys is tricky business so its not done very often.

Also, being that there are major changes in this area already I think this is the best time to make a change if you are going to.

Reply:
We changed the suffix for unique indexes in 4.3 already.

Reply:
You know, it's not even about the Forge. The point here is not to disrupt the consistency we have now. We may like the current constraint naming convention or not. It doesn't matter. The question is whether this change will do more harm than good. And IMO, it will not do any good for existing projects.

Saying that people are not using foreign keys much is arbitrary. But it may depend on the project we're working on.
Also, support in Forge is not an indicator of whether something is used. We can always use a simple query to run the SQL.

Sadly, not every change we want can be delivered immediately. Hundreds of developers use this framework, and their projects rely on the code we all deliver. Making changes is always tempting, but we need to remember the consequences.

Reply:
> We changed the suffix for unique indexes in 4.3 already.

Well... IMO it should not happen. I will search for the issue related to this.

Reply:
>You know, it's not even about the Forge.

I get your point but I'm thinking more of continuity between DBMS. If someone builds an app and want it to be compatible with different DBMS then things need to be consistent. Here the Oracle driver was added to the rest and the developer came across an issue specific to Oracle and decided to break the consistency.

Reply:
And I get your point, and it's valid. But... how often do we change the database engine during the project lifetime? Pretty close to never (yes, I took this data right from my ass 😅  ...sorry it was a long day).

I'm more focused on changes that potentially will make life harder for many developers than those that will cause problems for a few. 

Don't get me wrong. I want to see the changes you're proposing but not in a minor release.

Reply:
Not to keep arguing.. but imagine someone who develops an application to check and manage email. They might want to have their application play with a number of databases for people with different environments.

I'll revert the names back to the way they were.. but we keep the option to set our own names. This will at least make the tests simpler.

Reply:
> Not to keep arguing.

I don't feel like we argue. Seems like we are all on the same page. The discussion is about whether or not it's the right time to introduce some of these changes.

> but imagine someone who develops an application to check and manage email. They might want to have their application play with a number of databases for people with different environments.

Yes, that's a valid point. But to be fair, it's not that Oracle introduced this. We had an SQLite3 driver that behaved differently almost from the beginning.

The question is, what is the better approach for the existing projects? Having an additional line of code in the migration file to handle this situation or to force everyone to make changes in their databases?

For the minor release, I will choose the first option.

> I'll revert the names back to the way they were.. but we keep the option to set our own names. This will at least make the tests simpler.

Yes, the option to set our own names should stay.

Also... I know it might sound different, but I appreciate your contribution to the project and looking forward to more.

Reply:
Ok, got it reverted. All DBMS Except oracle use _foreign. Oracle uses _fk.